ludwig/models/llm.py

```diff
@@ -418,23 +489,117 @@ def get_args(self):
+        attention_masks = []
+        for i, merged_sample_ids in enumerate(merged_input_and_targets):
+            merged_input_and_targets[i] = self._add_left_padding(merged_sample_ids, max_length)
+            attention_masks.append(self._create_attention_mask(merged_input_and_targets[i]))
+
+        return torch.stack(merged_input_and_targets), torch.stack(attention_masks)
+
     def _remove_left_padding(self, input_ids_sample: torch.Tensor):
-        bos_idxs = torch.where(input_ids_sample == self.tokenizer.bos_token_id)[0]  # all BOS token locations
+        """Removes left padding from the input_ids tensor."""
+        # Remove all PAD tokens
+        pad_idxs = torch.where(input_ids_sample == self.tokenizer.pad_token_id)[0]  # all PAD token locations
+        if len(pad_idxs) != 0:
+            pad_idx = pad_idxs[-1]  # get last PAD token location
+        else:
+            pad_idx = 0
+        input_ids_sample_no_padding = input_ids_sample[pad_idx + 1 :]
+
+        # Start from the first BOS token
+        bos_idxs = torch.where(input_ids_sample_no_padding == self.tokenizer.bos_token_id)[0]  # all BOS token locations
         if len(bos_idxs) != 0:
             bos_idx = bos_idxs[0]  # get first BOS token location
         else:
             bos_idx = 0
 
-        input_ids_sample_no_padding = input_ids_sample[bos_idx:].unsqueeze(0)
-        return input_ids_sample_no_padding
+        input_ids_sample_no_bos = input_ids_sample_no_padding[bos_idx:].unsqueeze(0)
+        return input_ids_sample_no_bos
+
+    def _add_left_padding(self, input_ids, max_length, pad_value=0):
+        """Adds left padding to the input_ids tensor."""
+        if not pad_value:
+            padding = torch.zeros((max_length - input_ids.shape[0]), dtype=torch.int32, device=input_ids.device)
+        else:
+            padding = torch.tensor(
+                [pad_value] * (max_length - input_ids.shape[0]), dtype=torch.int32, device=input_ids.device
+            )
+        return torch.cat((padding, input_ids), dim=-1)
+
+    def _create_attention_mask(self, input_ids):
+        """Creates attention mask for the input_ids tensor."""
+        attention_mask = torch.ones_like(input_ids)
+        bos_index = (input_ids == self.tokenizer.bos_token_id).nonzero()
+
+        if len(bos_index) > 0:
+            first_bos_index = bos_index[0][0]
+            # Set attention mask to 0 for the part of the input that is padding
+            attention_mask[:first_bos_index] = 0
+
+        return attention_mask
 
     def get_augmentation_pipelines(self) -> AugmentationPipelines:
         """Returns the augmentation pipeline for this model."""
         return AugmentationPipelines({})
 
 
 def realign_target_and_prediction_tensors(
-    targets: Dict[str, torch.Tensor], predictions: Dict[str, torch.Tensor], of_name: str
+    targets: Dict[str, torch.Tensor],
+    predictions: Dict[str, torch.Tensor],
+    of_name: str,
+    tokenizer: AutoTokenizer,
```

Comment:
nit: tokenizer is probably actually of type BaseTokenizer or something along those lines